File: crates/store-db/src/punches.rs
//! Devices and the punches they push (CLAUDE.md §9).

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use sqlx::PgPool;
use store_core::identity::IdentityEvent;
use uuid::Uuid;

use crate::error::{FoundExt, Result};

/// A recorded punch.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct PunchRow {
    pub id: Uuid,
    pub device_id: Uuid,
    pub zk_user_id: String,
    pub device_ts: Option<DateTime<Utc>>,
    pub received_at: DateTime<Utc>,
    pub verify_mode: Option<String>,
    pub claimed: bool,
}

/// Register a device, or refresh what we know about one.
///
/// Called on every ADMS handshake, so a terminal that is swapped or reflashed
/// shows up without an admin having to add it first.
pub async fn upsert_device(
    pool: &PgPool,
    serial_no: &str,
    firmware: Option<&str>,
    timezone_offset_min: Option<i32>,
) -> Result<Uuid> {
    let id = sqlx::query_scalar!(
        r#"
        insert into devices (serial_no, name, last_seen_at, firmware, timezone_offset_min)
        values ($1, $1, now(), $2, $3)
        on conflict (serial_no) do update
           set last_seen_at = now(),
               firmware = coalesce(excluded.firmware, devices.firmware),
               timezone_offset_min = coalesce(
                 excluded.timezone_offset_min, devices.timezone_offset_min)
        returning id
        "#,
        serial_no,
        firmware,
        timezone_offset_min,
    )
    .fetch_one(pool)
    .await?;

    Ok(id)
}

pub async fn touch_device(pool: &PgPool, serial_no: &str) -> Result<Uuid> {
    upsert_device(pool, serial_no, None, None).await
}

/// Persist one punch, deduplicating retries (§9.1).
///
/// The device retries on any non-`OK` response, so the same batch arrives more
/// than once as a matter of course. `ON CONFLICT DO NOTHING` against the
/// `punches_dedup` index makes a retried batch a no-op rather than a second
/// punch.
///
/// Returns `(punch_id, is_new)`. `is_new` is false for a duplicate, which lets
/// the caller skip opening a second session without having to guess.
pub async fn record(
    pool: &PgPool,
    device_id: Uuid,
    event: &IdentityEvent,
) -> Result<(Uuid, bool)> {
    let inserted = sqlx::query_scalar!(
        r#"
        insert into punches (device_id, zk_user_id, device_ts, verify_mode, raw)
        values ($1, $2, $3, $4, $5)
        on conflict (device_id, zk_user_id, device_ts) do nothing
        returning id
        "#,
        device_id,
        event.external_user_id,
        event.device_ts,
        event.verify_mode.as_str(),
        event.raw,
    )
    .fetch_optional(pool)
    .await?;

    if let Some(id) = inserted {
        return Ok((id, true));
    }

    // Lost the race, or a genuine retry: fetch the row that is already there.
    let existing = sqlx::query_scalar!(
        r#"
        select id from punches
         where device_id = $1 and zk_user_id = $2 and device_ts is not distinct from $3
        "#,
        device_id,
        event.external_user_id,
        event.device_ts,
    )
    .fetch_optional(pool)
    .await?
    .or_not_found("punch")?;

    Ok((existing, false))
}

/// Punches whose `zk_user_id` matches no operator (§9.4).
///
/// Never dropped, always surfaced: an incomplete operator master is an admin
/// problem to fix, not a reason to lose the record that somebody opened the
/// door.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct UnknownUserPunch {
    pub punch_id: Uuid,
    pub zk_user_id: String,
    pub device_serial: String,
    pub received_at: DateTime<Utc>,
    pub occurrences: i64,
}

pub async fn unknown_user_punches(pool: &PgPool, limit: i64) -> Result<Vec<UnknownUserPunch>> {
    let rows = sqlx::query_as!(
        UnknownUserPunch,
        r#"
        select distinct on (p.zk_user_id)
               p.id          as "punch_id!",
               p.zk_user_id  as "zk_user_id!",
               d.serial_no   as "device_serial!",
               p.received_at as "received_at!",
               count(*) over (partition by p.zk_user_id) as "occurrences!"
          from punches p
          join devices d on d.id = p.device_id
         where not exists (
                 select 1 from operators o
                  where o.zk_user_id = p.zk_user_id and o.active)
         order by p.zk_user_id, p.received_at desc
         limit $1
        "#,
        limit.clamp(1, 500),
    )
    .fetch_all(pool)
    .await?;

    Ok(rows)
}

/// Recent punches, newest first — the device diagnostics view.
pub async fn recent(pool: &PgPool, limit: i64) -> Result<Vec<PunchRow>> {
    let rows = sqlx::query_as!(
        PunchRow,
        r#"
        select id           as "id!",
               device_id    as "device_id!",
               zk_user_id   as "zk_user_id!",
               device_ts,
               received_at  as "received_at!",
               verify_mode,
               claimed      as "claimed!"
          from punches
         order by received_at desc
         limit $1
        "#,
        limit.clamp(1, 1000),
    )
    .fetch_all(pool)
    .await?;

    Ok(rows)
}

/// When each device was last heard from — feeds `/admin/health` (§11).
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct DeviceHealth {
    pub id: Uuid,
    pub serial_no: String,
    pub name: Option<String>,
    pub location: Option<String>,
    pub last_seen_at: Option<DateTime<Utc>>,
    pub firmware: Option<String>,
}

pub async fn device_health(pool: &PgPool) -> Result<Vec<DeviceHealth>> {
    let rows = sqlx::query_as!(
        DeviceHealth,
        r#"
        select id as "id!", serial_no as "serial_no!", name, location,
               last_seen_at, firmware
          from devices
         order by serial_no
        "#
    )
    .fetch_all(pool)
    .await?;

    Ok(rows)
}

File: cloud/src/lib/punches.ts
// Persisting what the door terminal pushed, and turning it into a session
// offer (CLAUDE.md §9, §10).

import { sql } from "./db";
import type { AttlogRecord } from "./adms";

export type SessionOffer =
  | { kind: "OPENED"; sessionId: string; operatorId: string }
  | { kind: "ALREADY_OFFERED"; sessionId: string }
  | { kind: "UNKNOWN_OPERATOR" };

/** A device announces itself by serial; we have never met it before its first push. */
export async function upsertDevice(serial: string): Promise<string> {
  const rows = await sql<{ id: string }[]>`
    insert into devices (serial_no, last_seen_at)
    values (${serial}, now())
    on conflict (serial_no) do update set last_seen_at = now()
    returning id
  `;
  return rows[0].id;
}

/**
 * Record one punch.
 *
 * §9.1: the device retries on any non-OK response, so a retried batch must be a
 * no-op rather than a second punch. `punches_dedup` is what makes that true —
 * ON CONFLICT DO NOTHING returns no row for a duplicate, and the caller counts
 * it as accepted anyway, because from the device's point of view it was.
 *
 * §9.4: a punch whose user maps to no operator is still recorded. We never drop
 * data because the operator master is incomplete — that is how you end up
 * unable to explain a gap six months later.
 */
export async function recordPunch(
  deviceId: string,
  record: AttlogRecord,
): Promise<string | null> {
  const rows = await sql<{ id: string }[]>`
    insert into punches (device_id, zk_user_id, device_ts, verify_mode, raw)
    values (${deviceId}, ${record.userId}, ${record.deviceTs},
            ${record.verifyMode}, ${record.raw})
    on conflict do nothing
    returning id
  `;
  return rows[0]?.id ?? null;
}

/**
 * Offer a session for a punch.
 *
 * The `punch_id` unique constraint means a replayed punch cannot offer a second
 * session, so this is safe — and meaningful — to call more than once.
 */
export async function openFromPunch(punchId: string): Promise<SessionOffer> {
  return sql.begin(async (tx) => {
    const existing = await tx<{ id: string }[]>`
      select id from sessions where punch_id = ${punchId}
    `;
    if (existing[0]) {
      return { kind: "ALREADY_OFFERED", sessionId: existing[0].id } as SessionOffer;
    }

    const operator = await tx<{ id: string }[]>`
      select o.id from punches p
        join operators o on o.zk_user_id = p.zk_user_id and o.active
       where p.id = ${punchId}
    `;
    if (!operator[0]) {
      return { kind: "UNKNOWN_OPERATOR" } as SessionOffer;
    }

    const session = await tx<{ id: string }[]>`
      insert into sessions (operator_id, punch_id, state, manual_identity)
      values (${operator[0].id}, ${punchId}, 'UNCLAIMED', false)
      returning id
    `;

    await tx`update punches set claimed = true where id = ${punchId}`;

    return {
      kind: "OPENED",
      sessionId: session[0].id,
      operatorId: operator[0].id,
    } as SessionOffer;
  }) as Promise<SessionOffer>;
}
